File: src/main/java/com/kiseru/asteroids/server/logics/Screen.java
package com.kiseru.asteroids.server.logics;

import com.kiseru.asteroids.server.logics.auxiliary.Coordinates;
import com.kiseru.asteroids.server.logics.models.Point;
import java.util.Arrays;

/**
 * @author Bulat Giniyatullin
 * 09 Декабрь 2017
 */

public class Screen {
    private final int width;
    private final int height;
    private final String[][] mainMatrix;

    public Screen(int width, int height) {
        this.width = width;
        this.height = height;
        this.mainMatrix = new String[height + 2][width + 2];
        generateClearScreen();
    }

    public void render(Point point) {
        if (point.isVisible()) {
            draw(point.getCoordinates(), point.view());
        }
    }

    /**
     * рисует на экране точку
     * @param coordinates - по какой координате
     * @param symbol - символ, которым отображается точка
     */
    private void draw(Coordinates coordinates, String symbol) {
        if (symbol == null || symbol.isBlank()) {
            mainMatrix[coordinates.y()][coordinates.x()] = ".";
        }

        if (mainMatrix[coordinates.y()][coordinates.x()].equals(".")) {
            mainMatrix[coordinates.y()][coordinates.x()] = symbol;
        } else {
            mainMatrix[coordinates.y()][coordinates.x()] =
                    String.format("%s|%s", mainMatrix[coordinates.y()][coordinates.x()], symbol);
        }
    }

    /**
     * обновляет экран
     */
    public void update() {
        generateClearScreen();
    }

    /**
     * отображает экран
     */
    public String display() {
        StringBuilder result = new StringBuilder();
        for (int i = 1; i < height + 1; i++) {
            for (int j = 1; j < width + 1; j++) {
                result.append(mainMatrix[i][j]);
                result.append("\t");
            }
            result.append("\n");
        }
        return result.toString();
    }

    private void generateClearScreen() {
        for (int i = 0; i < height + 2; i++) {
            if (i == 0 || i == height + 1) {
                Arrays.fill(mainMatrix[i], "*");
            } else {
                Arrays.fill(mainMatrix[i], ".");
                mainMatrix[i][0] = "*";
                mainMatrix[i][width + 1] = "*";
            }
        }
    }

    public int getHeight() {
        return height;
    }

    public int getWidth() {
        return width;
    }
}

File: src/main/kotlin/com/kiseru/asteroids/server/model/Screen.kt
package com.kiseru.asteroids.server.model

import java.util.Arrays

class Screen(
    val width: Int,
    val height: Int,
) {

    private val mainMatrix = Array(height + 2) { Array(width + 2) { "." } }

    init {
        generateClearScreen()
    }

    fun render(gameObject: GameObject) {
        if (gameObject.isVisible) {
            draw(gameObject)
        }
    }

    private fun draw(gameObject: GameObject): Unit =
        draw(gameObject.x, gameObject.y, gameObject.view())

    private fun draw(x: Int, y: Int, symbol: String) {
        if (symbol.isBlank()) {
            mainMatrix[y][x] = "."
        }

        if (mainMatrix[y][x] == ".") {
            mainMatrix[y][x] = symbol
        } else {
            mainMatrix[y][x] = "${mainMatrix[y][x]}|$symbol"
        }
    }

    fun update() {
        generateClearScreen()
    }

    private fun generateClearScreen() {
        for (i in 0 until height + 2) {
            if (i == 0 || i == height + 1) {
                Arrays.fill(mainMatrix[i], "*")
            } else {
                Arrays.fill(mainMatrix[i], ".")
                mainMatrix[i][0] = "*"
                mainMatrix[i][width + 1] = "*"
            }
        }
    }

    fun display(): String {
        val stringBuilder = StringBuilder()
        for (i in 1 until height + 1) {
            for (j in 1 until width + 1) {
                stringBuilder.append(mainMatrix[i][j])
                stringBuilder.append("\t")
            }
            stringBuilder.append("\n")
        }
        return stringBuilder.toString()
    }
}